Authenticated Images

Allcharles opened the pull request:
# Authenticated Images

Create authenticated image requests and handle defaults for model images.

## Changes

- Created `AuthenticatedImageDirective`
- Created `BawImage` decorator
- Update app `<img />` elements to use new directive
- Added ng-mocks library (https://github.com/ike18t/ng-mocks)
- Rewrote all CardModule component unit tests using spectator

## Issues

- [x] Need to append authentication token to urls
  - Append to urls which contain the `API_ROOT` in their path only
  - This should be handled by the `ImageDirective`
  - Add input to `ImageDirective` which disables adding authentication token to urls
- [x] Need to rename `ImageDirective` and `ImageDirectiveModule` to more descriptive names
- [x] Tests are failing
- [x] Need to write tests for decorator and directive
- [x] Re-write card component tests to jasmine
- [x] Create default image url if none provided for `BawImage` directive
- [x] Edit `assertImage` function to assert that images are using the `ImageDirective`
- [x] Rebase branch

## Closes

Closes #232

Allcharles committed:
Created BawImage decorator

Allcharles committed:
Created image directive

Allcharles committed:
Added image directive to images

- Created asset root environment variable
- Added BawImage decorator to models

Allcharles committed:
Fix card-image component

Allcharles committed:
Cleanup image directive

- Changed selector to target all img elements
- Cleanup code structure

Allcharles committed:
Added authTokens to urls

Allcharles committed:
Rename imageDirective to AuthenticatedImageDirective

Allcharles committed:
Cleanup broken tests

- Updated assertImage to detect authenticated images
- Cleanup basic tests

Allcharles committed:
Cleanup basic card unit tests

Allcharles committed:
Cleanup remaining cards unit tests

Allcharles committed:
Added ng-mocks dep

Allcharles committed:
Created placeholder unit tests

Allcharles committed:
Implemented image directive unit tests

- Implemented ImageDirective unit tests
- Renamed disableAuthentication to disableAuth
- Created 404 image

Allcharles committed:
Improve accuracy of test for AuthenticatedImageDirective

Allcharles committed:
Implemented BawImage unit tests

Allcharles committed:
Merge branch 'master' into image-directive

atruskie commented on src/app/directives/image/image.directive.spec.ts:
```diff
@@ -0,0 +1,246 @@
+      const image = getImage();
+      [1, 2, 3].forEach(() => createImgErrorEvent(image));
+      assertImage(image, imageUrls[3].url, "alt");
+    });
```
very minor nit: using more descriptive test names would act like documentation.

Instead of `should handle url error` maybe: `given an invalid url, it loads the next available url`

and:
 `given multiple bad urls, it tries to load each until one succeeds`

atruskie commented on src/app/directives/image/image.directive.spec.ts:
```diff
@@ -0,0 +1,246 @@
+      const imageUrls = modelData.imageUrls();
+      spectator = createThumbnailDirective(imageUrls, ImageSizes.MEDIUM);
+      // MEDIUM image size is created at 2nd index by modelData.imageUrls
+      assertImage(getImage(), imageUrls[2].url, "alt");
```
Don't hard code this index. Try `imageUrls.filter(x => x.size == imageSizes.MEDIUM).url`

atruskie commented on src/app/directives/image/image.directive.ts:
```diff
@@ -0,0 +1,136 @@
+export const image404RelativeSrc = "/assets/images/404.png";
+
+@Directive({
+  // tslint:disable-next-line: directive-selector
```
explain why we think it is ok to disable this warning

atruskie commented on src/app/directives/image/image.directive.ts:
```diff
@@ -0,0 +1,136 @@
+    // On Component Initial Load
+    if (changes.src.isFirstChange()) {
+      // Set attribute so that unit tests can track authenticated images
+      this.imageRef.nativeElement.setAttribute("bawImage", "true");
```
Any other solution? This just seems like noise

atruskie commented on src/app/directives/image/image.directive.ts:
```diff
@@ -0,0 +1,136 @@
+  ) {}
+
+  ngOnChanges(changes: SimpleChanges): void {
+    // On Component Initial Load
```
you should reset `srcIndex` if the input changes - it might be a new array, differently sized, or with different valid/invalid urls

atruskie commented on src/app/directives/image/image.directive.ts:
```diff
@@ -0,0 +1,136 @@
+  /**
+   * Tracks which src value to display, unless thumbnail is used
+   */
+  private srcIndex = 0;
```
another way you could handle tracking bad images would be to:
- tag the image object/url when it's faulty
- or keep a `Set<>` of URLs

Either way might make it easier to track bad urls. See comment about resetting index below

atruskie commented on src/app/directives/image/image.directive.ts:
```diff
@@ -0,0 +1,136 @@
+  /**
+   * Retrieve thumbnail image url or return null
+   */
+  private retrieveThumbnailIfExists(): string | null {
```
could probably express this with a `.filter` for succinctness

atruskie reviewed (changes requested):
Excellent work. I'd say the specs for the image directive need a little more descriptive documentation - as per the comment I think explaining what the test is testing in the test title would be enough.

Also ( i know we've talked about this before) but why not have the model decorator append the auth token?

Allcharles committed:
Address review comments

Commit addresses most review comments. Need to solve edge case in image directive where thumbnails are not used last in new solution.

Allcharles committed:
Better default image handling

Implemented a better method for default image handling. Also implemented a new system where newly added src urls are added to the front of the cue.

Allcharles committed:
Remove bawImage attribute from image directive

Allcharles committed:
Merge branch 'master' into image-directive

Allcharles committed:
Update dependencies

Allcharles committed:
Fix CI tests

Allcharles committed:
Separate browser test suites in workflow
